kirkwood: expand internal storage for EA3500/EA4500 & external for all

builker opened the pull request:
1. Increase EA3500/EA4500's kernel size to 3072 KB, and rootfs to the rest.
2. Restore booting support for all Kirkwood devices from external devices,
   which is specially needed for EA3500/EA4500's flash/upgrade to 21.02.

Signed-off-by: Eubert Bao <bunnier@gmail.com>

```
root@EA3500:/# cat /proc/mtd
dev:    size   erasesize  name
mtd0: 00080000 00004000 "u-boot"
mtd1: 00004000 00004000 "u_env"
mtd2: 00004000 00004000 "s_env"
mtd3: 00300000 00004000 "kernel"
mtd4: 03b00000 00004000 "ubi"
mtd5: 00178000 00004000 "unused"
root@EA3500:/# df
Filesystem           1K-blocks      Used Available Use% Mounted on
/dev/root                 2816      2816         0 100% /rom
tmpfs                    27892        52     27840   0% /tmp
/dev/ubi0_1              40092        40     37488   0% /overlay
overlayfs:/overlay       40092        40     37488   0% /
tmpfs                      512         0       512   0% /dev
```
```
root@EA4500:/# cat /proc/mtd
dev:    size   erasesize  name
mtd0: 00080000 00020000 "u-boot"
mtd1: 00020000 00020000 "u_env"
mtd2: 00020000 00020000 "s_env"
mtd3: 00300000 00020000 "kernel"
mtd4: 07b00000 00020000 "ubi"
mtd5: 00140000 00020000 "unused"
root@EA4500:/# df
Filesystem           1K-blocks      Used Available Use% Mounted on
/dev/root                 2816      2816         0 100% /rom
tmpfs                    60548        56     60492   0% /tmp
/dev/ubi0_1             109264        40    104388   0% /overlay
overlayfs:/overlay      109264        40    104388   0% /
tmpfs                      512         0       512   0% /dev
```

builker committed:
kirkwood: expand nand storage for EA3500/EA4500 & external for all
1. Increase EA3500/EA4500's kernel size to 3072 KB, and rootfs to the rest.
2. Restore booting support for all Kirkwood devices from external devices,
   which is specially needed for EA3500/EA4500's flash/upgrade to 21.02.

Signed-off-by: Eubert Bao <bunnier@gmail.com>

bobafetthotmail commented:
-- I don't like losing the dual firmware setup, I would prefer if you just enlarged the kernel size by taking the space from its firmware partition. Also is this really necessary? I thought the kernel shrunk by quite a bit by removing all the usb/sata/mmc support.

--Imho re-adding a fully-fledged boot from storage is not necessary for the upgrade (nor any of the other devices need this).
You can just set uboot to load and boot a initramfs image like **linksys_ea4500-initramfs-uImage** from usb (this is a full OpenWrt image with the same packages as the sysupgrade image, but it is fully loaded to RAM by the uboot and then runs from RAM so it does not need storage drivers to load things from USB/Sata/MMC) and then connect to its web interface or ssh at 192.168.1.1 and do a sysupgrade from it to re-partition and install.
an example of the commands to boot an initramfs image from a normal FAT32 flash drive 
**usb reset ; fatload usb 0:1 0x2000000 linksys_ea4500-initramfs-uImage ; bootm 0x2000000 ;** 

--also imho this upgrade procedure should not be automated like that but should just print an error message and point to the wiki for an upgrade tutorial, where you will write the steps to do this and the users can have more information on what is happening and why. What you did here is dangerous, asking the user to edit uboot like that out of the blue will break luci sysupgrade and confuse the users.

--For convenience the uboot command to load the initramfs image should be fatload, not ext2load, so you can load the initramfs image from a normal USB drive formatted as FAT32, which is doable for a Windows system. Requiring to format as ext2-3-4  will be a pain for most users, which will have a Windows or Mac. Chech that the uboot actually has that, afaik it should.

--This commit is incomplete, you did not add ext4 images to the build system. How is Openwrt supposed to boot from external storage if you don't make an image for external storage? Do you require the user to figure out how to prepare such images manually? Where are the instructions for that? As said above I don't think making a fully fledged external device boot is required just to do the upgrade, you can do it fine with a sysupgrade image we build already and does not need usb/sata/mmc support in kernel.

adschm commented:
Hi, commit title is too long, and both "kirkwood:" and "expand" should be lower case.

builker commented:
Thank you, Alberto, for your comments. Allow me to present [the problem](https://forum.openwrt.org/t/kirkwood-devices-are-limited-by-ea3500-ea4500-kernel-size-how-to-increase-it/90736), which should have been done before the solution above.

[The external storage/USB booting is no longer supported for all the Kirkwood devices](https://forum.openwrt.org/t/ea3500-snapshots-of-the-master-branch-get-endless-boot-loop/87612), [as external storage code was moved from kernel to modules on the master branch in June, 2020](https://github.com/openwrt/openwrt/commits/master/target/linux/kirkwood), because their kernel size hit the limit imposed by the Linksys stock layout (dual firmware setup):
- EA3500: 2624k
- EA4500: 2688k
They are the smallest 2 targets in the Kirkwood system, but eliminate the big possibility to GBs and TBs 2 for all the targets. **It's not fair** :sob:. Ironically, both have quite bit of space, 64/128 MB in total.

The dual firmware setup is a luxury that the majority of OpenWrt flashed targets don't have. I liked the stock firmware as a backup for all my 10 production routers, 8 of them in different retail centers. In 2~3 years, Linksys stock firmware was never booted/activated. OpenWrt has been too stable. Is there is a feature in the stock firmware that OpenWrt doesn't offer? If yes, [how about a separate EA3500](https://www.amazon.com/dp/B00L4HWY3E)?

Trading away the luxury gains ~30/80 MB on 3500/4500 internal nand, enabling tens of features, such as code heavy VoIP and mail servers, that OpenWrt has to offer, without setting up exroot. [Other users asked for more internal space.](https://forum.openwrt.org/t/linksys-ea3500-64mb-rom-only-9mb-free-after-lede/10825)

Not sure `fatload` is supported by the version of U-Boot that comes with EA3500/EA4500. Can you confirm either way?

The following flash upgrade procedure should be documented for 21.02:
1. flash OpenWrt 19.07.x factory.bin via stock's Web UI
2. in 19.07.x, run 21.02 version of sysupgrade-enable-usb-boot
3. boot up 21.02 uImage from a USB drive or U-Boot via a serial cable, and sysupgrade sysupgrade.bin

Making a bootable USB drive (ext2/3/4 formatted) can be done in a Linux environment or OpenWrt:
`root@EA4500# mkdir /boot; cp openwrt-snapshot-kirkwood-linksys_ea4500-initramfs-uImage /boot/kernel`
That's it, a rescue/upgrade USB drive is made. I am afraid an ex4 image overwrites a USB drive. [A bootable drive with `rootfs` takes untar sysupgrade.bin, cp, unsquashfs without touching its `/home` directory.](https://forum.openwrt.org/t/linksys-ea3500-ea4500-tricks-how-tos-issues/34510/2?u=builder)

Hope the above addressed your concerns. Any other issues?

bobafetthotmail commented:
> Thank you, Alberto, for your comments. Allow me to present [the problem](https://forum.openwrt.org/t/kirkwood-devices-are-limited-by-ea3500-ea4500-kernel-size-how-to-increase-it/90736), which should have been done before the solution above.

Most of what you told me here should be added in the commit description. Like why you are doing this, and instructions to use it.
As stated in the submission rules https://openwrt.org/submitting-patches

> 
> [The external storage/USB booting is no longer supported for all the Kirkwood devices](https://forum.openwrt.org/t/ea3500-snapshots-of-the-master-branch-get-endless-boot-loop/87612), as [external storage code was moved from kernel to modules on the master branch in June, 2020](https://github.com/openwrt/openwrt/commits/master/target/linux/kirkwood), because their kernel size hit the limit imposed by the Linksys stock layout (dual firmware setup):

Ok, you want USB for this device but there is no need for Sata and MMC too. Just add USB in kernel. You increased kernel partition size a bit but it's still not big, and no other Kirkwood device needs that so it's still good to not waste kernel space.

Also, I think you are joining two different changes here, one is the "increase partition size in EA3500/EA4500 by dropping dual firmware support"  and another is "allow boot from USB for all Kirkwood", it is best to split them into different commits as they are two different things with not much in common. As far as I know you can even make a different PR with the "boot from USB" changes as none of the changes in it affect the "increase partition size" commit.
Again this is also from the contribution rules and probably the next thing **adschm** will also tell you.

I still don't understand the statement of "USB boot is needed for firmware upgrade to 21.02" in commit or comments in the scripts/makefile, it's not needed for upgrade to 21.02 firmware and it is also not needed for the partition layout change.

As I said the only thing needed is to load an initramfs image like **linksys_ea4500-initramfs-uImage** from uboot, and do sysupgrade from there to install with a different partition layout.

So the only thing you need for that is to add usb boot configuration to the uboot, with the script you added or with some instructions in the wiki.
**usb reset ; fatload usb 0:1 0x2000000 linksys_ea4500-initramfs-uImage ; bootm 0x2000000 ;** will load and boot such initramfs image from a FAT32 flash drive.

> The dual firmware setup is a luxury that the majority of OpenWrt flashed targets don't have. I liked the stock firmware as a backup for all my 10 production routers, 8 of them in different retail centers. In 2~3 years, Linksys stock firmware was never booted/activated.

It's actually used for OpenWrt, it's not just dead space with the stock firmware.

See these two files here https://github.com/openwrt/openwrt/tree/master/target/linux/kirkwood/base-files/lib/upgrade
When you sysupgrade it will write to secondary partition and then adjust uboot config to boot from it.
I see that you did not touch these files in your commit. I think this will be a problem when doing a sysupgrade.

I have used this feature plenty and it has been useful for me for snapshot as sometimes it will fail to boot after sysupgrade and with this system it will boot back to the other partition.
And this is why I ask you if you can please not drop this functionality, it is one of the few devices where it is actually implemented in OpenWrt.

> Trading away the luxury gains ~30/80 MB on 3500/4500 internal nand, enabling tens of features, such as code heavy VoIP and mail servers, that OpenWrt has to offer, without setting up exroot. [Other users asked for more internal space.](https://forum.openwrt.org/t/linksys-ea3500-64mb-rom-only-9mb-free-after-lede/10825)

You can reclaim wasted space in "syscfg" partition without nuking everything like that, OpenWrt has a driver called **mtdconcat** to "join" different partitions into a single one to handle situations like this where you want to use flash from a non-contiguous partition but can't (or don't want) to actually erase the partitions between them.

see here for the patch and the commit with the instructions https://github.com/openwrt/openwrt/blob/b10d6044599d8c1fa7fbb2374bcbf30118d39db1/target/linux/generic/pending-5.10/497-mtd-mtdconcat-add-dt-driver-for-concat-devices.patch
And here a commit that is using this driver in a device to reclaim space from other unused partitions, it should be easy enough to do the same https://github.com/openwrt/openwrt/commit/13581a5fa7ed09aa04393d496f44d0bf068e3024

So you can split the syscfg in two 8Mb partitions and add one of them to each of the root partitions.
This doubles current root partition size, so it's still worth it even if it is not as good as taking all the flash space like you are doing here.

> 
> Not sure `fatload` is supported by the version of U-Boot that comes with EA3500/EA4500. Can you confirm either way?
> 

Ok, I have serial console so I can talk to uboot console directly

fatload is present and in my test it also works correctly with the string of commands I mentioned above.

Note that if you want to boot from a ext2/4 rootfs in a usb you still need to use ext2load. You can just chain the commands so that first it tries with fatload for the initramfs image, then ext2load for the kernel in an ext2/4 partition, and if both fail it will try loading from internal flash.


These are all commands available in the uboot of EA5400 (and afaik it's the same uboot of the other device, and the same for most kirkwoods too for that matter)
```
Viper>> help
?       - alias for 'help'
SatR - sample at reset sub-system, relevent for DB only
base    - print or set address offset
boot    - boot default, i.e., run 'bootcmd'
bootd   - boot default, i.e., run 'bootcmd'
bootext2    dev:boot_part1,boot_part2 addr boot_image linux_dev_name 
bootm   - boot application image from memory
bootp	- boot image via network using BootP/TFTP protocol
bubt	- Burn an image on the Boot Nand Flash.
chpart	- change active partition
cmp     - memory compare
cmpm	- Compare Memory
cp      - memory copy
cpumap - Display CPU memory mapping settings.
crc32   - checksum calculation
date    - get/set/reset date & time
dclk	- Display the MV device CLKs.
dhcp	- invoke DHCP client to obtain IP/boot params
diskboot- boot from IDE device
echo    - echo args to console
eeprom  - EEPROM sub-system
erase   - erase FLASH memory
ext2load- load binary file from a Ext2 filesystem
ext2ls  - list files in a directory (default /)
fatinfo - print information about filesystem
fatload - load binary file from a dos filesystem
fatls   - list files in a directory (default /)
fi	- Find value in the memory.
flinfo  - print FLASH memory information
fsinfo	- print information about filesystems
fsload	- load binary file from a filesystem image
g	- start application at cached address 'addr'(default addr 0x40000)
go      - start application at address 'addr'
help    - print online help
icrc32  - checksum calculation
ide     - IDE sub-system
iloop   - infinite loop on address range
imd     - i2c memory display
iminfo  - print header information for application image
imm[.b, .s, .w, .l]     - i2c memory modify (auto-incrementing)
imw     - memory write (fill)
inm     - memory modify (constant address)
iprobe  - probe to discover valid I2C chip addresses
ir	- reading and changing MV internal register values.
loop    - infinite loop on address range
ls	- list files in a directory (default /)
map	- Diasplay address decode windows
md      - memory display
me	- PCI master enable
mfg	- Diagnostic tools for MFG
mm      - memory modify (auto-incrementing)
mmcinit - init mmc card
mp	- map PCI BAR
mtdparts- define flash/nand partitions
mtest   - simple RAM test
mw      - memory write (fill)
nand                   - NAND sub-system
nboot   - boot from NAND device
nbubt	- Burn a boot loader image on the Boot Nand Flash.
nm      - memory modify (constant address)
pci     - list and access PCI Configuration Space
phyRead	- Read PCI-E Phy register
pciePhyWrite	- Write PCI-E Phy register
phyRead	- Read Phy register
phyWrite	- Write Phy register
ping	- send ICMP ECHO_REQUEST to network host
printenv- print environment variables
protect - enable or disable FLASH write protection
rarpboot- boot image via network using RARP/TFTP protocol
rcvr	- Satrt recovery process (Distress Beacon with TFTP server)
reset   - Perform RESET of the CPU
resetenv	- Return all environment variable to default.
run     - run commands in an environment variable
saveenv - save environment variables to persistent storage
se	- PCI Slave enable
setenv  - set environment variables
sflash	- read, write or erase the external SPI Flash.
sg	- scanning the PHYs status
sp	- Scan PCI bus.
switchCountersRead	- Read switch port counters.
switchRegRead	- Read switch register
switchRegWrite	- Write switch register
Temp	- read chip Tj temp
tftpboot- boot image via network using TFTP protocol
usb     - USB sub-system
usbboot - boot from USB device
version - print monitor version
```


> The following flash upgrade procedure should be documented for 21.02:
> 
>     1. flash OpenWrt 19.07.x factory.bin via stock's Web UI
> 
>     2. in 19.07.x, run 21.02 version of sysupgrade-enable-usb-boot
> 
>     3. boot up 21.02 uImage from a USB drive or U-Boot via a serial cable, and sysupgrade sysupgrade.bin
> 

as mentioned above, this should be written in the commit description at the very least

> 
> Making a bootable USB drive (ext2/3/4 formatted) can be done in a Linux environment or OpenWrt:
> `root@EA4500# mkdir /boot; cp openwrt-snapshot-kirkwood-linksys_ea4500-initramfs-uImage /boot/kernel`
> That's it, a rescue/upgrade USB drive is made. I am afraid an ex4 image overwrites a USB drive.
 [A bootable drive with `rootfs` takes untar sysupgrade.bin, cp, unsquashfs without touching its `/home` directory.](https://forum.openwrt.org/t/linksys-ea3500-ea4500-tricks-how-tos-issues/34510/2?u=builder)
> 

And this should also go in commit description, for the commit where you add USB boot support.


> Hope the above addressed your concerns. Any other issues?

Yes. You are making device-bricking changes without also adding a safety in sysupgrade procedure. If someone does a sysupgrade on a normal system without doing the special migration step with the initramfs first, he will brick the device. This should be avoided.

Edit the files in https://github.com/openwrt/openwrt/tree/master/target/linux/kirkwood/base-files/lib/upgrade and at the very least add an error message, IMHO you should also make sysupgrade automatically fail if it detects the "old partition layout" unless the system is also running from RAM (i.e. it is an initramfs image).
You should make sure nobody can do a sysupgrade that will brick his device.

builker committed:
kirkwood: bump up Linksys DEVICE_COMPAT_VERSION to 2.0
Add DEVICE_COMPAT_MESSAGE stating the change of partition layout,
upgrade options, and a warning against force upgrade.

Signed-off-by: Eubert Bao <bunnier@gmail.com>

builker commented:
It's quite time and energy consuming to read lengthy and repetitive comments. Unlike Liniksys stock firmware, personal preferences can be customized in OpenWrt. The scope of is PR is to (1) continue the capability to boot from external storage for **all** Kirkwood devices, supported in release 19.07; (2) remove the bottleneck, EA3500/EA4500's stock `kernel` size; (3) rather shrinking already small EA3500's `rootfs` partition, [enlarge it to what folks have expected and asked for](https://forum.openwrt.org/t/linksys-ea3500-64mb-rom-only-9mb-free-after-lede/10825). The 3 Linksys devices will finally join the mainstream of Kirkwood and other OpenWrt devices.

The question on the migration 19.07 to 21.02 prompts me to think about it more. 3 Linksys devices' DEVICE_COMPAT_VERSION is already bumped to 1.1 due to DSA migration:
```
 -----------------------------------------------------
OpenWrt 19.07.6, r11278-8055e38794
 -----------------------------------------------------
root@EA3500:~# sysupgrade /tmp/openwrt-snapshot-kirkwood-linksys_ea3500-squashfs-sysupgrade.bin
Device linksys,audi not supported by this image
Supported devices: linksys,ea3500 linksys,audi linksys-audi - Image version mismatch: image 1.1, device 1.0. Please wipe config during upgrade (force required) or reinstall. Reason: Config cannot be migrated from swconfig to DSA
Image check failed.
```
It doesn't sound serious enough. Giving the core developers a choice, I add a commit spelling out the danger, [similar to mvebu/cortexa9](https://github.com/openwrt/openwrt/blob/master/target/linux/mvebu/image/cortexa9.mk):
```
root@EA3500:/# sysupgrade /tmp/openwrt-snapshot-kirkwood-linksys_ea3500-squashfs-sysupgrade.bin
Device linksys,audi not supported by this image
Supported devices: linksys,ea3500 linksys,audi linksys-audi - Image version mismatch: image 2.0, device 1.0. Please wipe config during upgrade (force required) or reinstall. Reason: Config cannot be migrated from swconfig to DSA. Partition layout has changed compared with older versions (up to 19.07) due to kernel size restrictions. Sysupgrade needs to be done in a later version than 19.07, which can be booted from a USB drive. Alternatively, perform a new installation via TFTP in the U-Boot. For instructions, see your device's page on https://openwrt.org/toh. Warning: Force upgrade (sysupgrade -F | --force) will likely brick this device!
Image check failed.
```
Thanks for helping make this PR better.

adschm commented on target/linux/kirkwood/base-files/sbin/sysupgrade-enable-usb-boot:
```diff
@@ -0,0 +1,45 @@
+#!/bin/sh
```
All new files should have an SPDX license identifier.

adschm commented on target/linux/kirkwood/base-files/sbin/sysupgrade-enable-usb-boot:
```diff
@@ -0,0 +1,45 @@
+         openwrt_usb_boot  usb start; mw \${loadaddr} 0 8; ext2load usb 0:1 \${loadaddr} /boot/kernel; setenv bootargs \${console} \${mtdparts} \${usb_fs_bootargs_root} rootdelay=10; bootm \${loadaddr}; usb stop
+EOF
+
+    echo "Enable OpenWrt USB boot.  Changing the following u-boot varilables:"
```
Typo.

adschm commented on target/linux/kirkwood/base-files/sbin/sysupgrade-enable-usb-boot:
```diff
@@ -0,0 +1,45 @@
+    linksys,audi|\
+    linksys,e4200-v2|\
+    linksys,ea4500|\
+    linksys,viper)
```
I don't think we use linksys,audi or linksys,viper anywhere.

adschm commented on target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-ea3500.dts:
```diff
@@ -164,28 +164,13 @@
-			reg = <0x2a00000  0x1600000>;
+		partition@800000 {
+			label = "rootfs";
+			reg = <0x500000 0x3B00000>;
```
Please use lower case for hex characters.

adschm commented on target/linux/kirkwood/image/Makefile:
```diff
@@ -94,19 +91,24 @@ endef
+  # nandboot in file base-files/sbin/sysupgrade-enable-usb-boot:
+  #   "nand read.e 0x2000000 0x200000 0x300000", the last value in bytes.
+  #
+  # Note: The support for USB booting is needed for sysupgrade to 21.02.
```
This is not part of 21.02.

adschm commented on target/linux/kirkwood/image/Makefile:
```diff
@@ -102,11 +97,19 @@ define Device/linksys
+	Alternatively, perform a new installation via TFTP in the U-Boot. \
+	For instructions, see your device's page on https://openwrt.org/toh. \
+	Warning: Force upgrade (sysupgrade -F | --force) will likely brick \
+	this device!
```
DSA comment would be irrelevant now.

And I don't see why you would add this here, where all Linksys device are and will be affected, even new ones.

adschm commented:
@bobafetthotmail 's comments are generally valid and appropriate.

@builker You still don't explain and justify your changes in the commit message.
I don't think dropping the OEM partitions is _necessary_ (yet).
I don't think introducing a new, additional sysupgrade mechanism is a good idea.
I don't think your attitude when responding is appropriate.

This looks to me like a custom solution for a custom problem. Do we really need it for our platform default setup?

arrmo commented:
Hi,

I may be missing what you are after - but I found this because I was having the same issue (wanted to increase the kernel size ... trying testing kernel, but it was too large => build failed). I made just a couple minor changes,
1) change the dts, as @bobafetthotmail seems to be saying. This is to shift as bit of the flash space from rootfs to kernel.
2) with1) done, then change the makefile, to match the kernel size limit

The outcome is that I now am running v5.10 kernel on this router, with the partitions like this,
```
dev:    size   erasesize  name
mtd0: 00080000 00004000 "u-boot"
mtd1: 00004000 00004000 "u_env"
mtd2: 00004000 00004000 "s_env"
mtd3: 00300000 00004000 "kernel1"
mtd4: 01100000 00004000 "ubi"
mtd5: 00300000 00004000 "kernel2"
mtd6: 01100000 00004000 "rootfs2"
mtd7: 01600000 00004000 "syscfg"
mtd8: 00178000 00004000 "unused"
```

I may be missing the point, but is this what you are after?

@bobafetthotmail, do you also want to try this on your 4500? I can share my changes if you want. And yes, as you note above - I didn't lose dual partition (I like that too!).

Thanks!


PS; I don't think the info under the EA3500 is quite right - meaning this,
`  3. Repeat the top steps three times, and when the power light no longer turns off during boot up, the device should be reset to default firmware/ settings.`
=> Rather, this sends you back to the "other" partition ... has saved my butt a few times 😆.

arrmo commented:
Actually, thinking about this one, but ... does the larger kernel (partitioning change) need to be a new recipe (device)? Asking because it's sort of a two step process to get there => separate that out, to ease transitioning?

What I mean is ... change the partitioning, "old" and "new" device, but only allow the larger kernel in the "new" device, to make sure it's not done by accident.

Thoughts?

Thanks!
